### ISSUE TITLE
Indexing fails on surrogates

### ISSUE THREAD

**Issue by CollinScripter** [mxmlnkn/ratarmount#32]
Trying to open a file but it seems to error at one point due to not being able to utf-8 encode surrogates. I made a new tar containing surrogates in file names to try to reproduce but was unable to. Unfortunately I cannot open the tar using other programs to see any culprit files, as it is ~700GB.
Taceback:
```
Traceback (most recent call last):
  File "/home/collin/.local/bin/ratarmount", line 8, in <module>
    sys.exit(cli())
  File "/home/collin/.local/lib/python3.8/site-packages/ratarmount.py", line 1557, in cli
    fuseOperationsObject = TarMount(
  File "/home/collin/.local/lib/python3.8/site-packages/ratarmount.py", line 995, in __init__
    self.mountSources = [ self._openTar( tarFile, clearIndexCache, recursive, gzipSeekPointSpacing )
  File "/home/collin/.local/lib/python3.8/site-packages/ratarmount.py", line 995, in <listcomp>
    self.mountSources = [ self._openTar( tarFile, clearIndexCache, recursive, gzipSeekPointSpacing )
  File "/home/collin/.local/lib/python3.8/site-packages/ratarmount.py", line 1039, in _openTar
    return SQLiteIndexedTar( tarFilePath,
  File "/home/collin/.local/lib/python3.8/site-packages/ratarmount.py", line 294, in __init__
    self.createIndex( self.tarFileObject )
  File "/home/collin/.local/lib/python3.8/site-packages/ratarmount.py", line 479, in createIndex
    self._setFileInfo( fileInfo )
  File "/home/collin/.local/lib/python3.8/site-packages/ratarmount.py", line 657, in _setFileInfo
    self.sqlConnection.execute( 'INSERT OR REPLACE INTO "files" VALUES (' +
UnicodeEncodeError: 'utf-8' codec can't encode character '\udc83' in position 2: surrogates not allowed
Exception ignored in: <function TarMount.__del__ at 0x7fe07c99d550>
Traceback (most recent call last):
  File "/home/collin/.local/lib/python3.8/site-packages/ratarmount.py", line 1036, in __del__
    os.close( self.mountPointFd )
AttributeError: mountPointFd
```

**Comment by mxmlnkn** [comment]
In short, it's an [encoding issue](https://docs.python.org/3/library/tarfile.html#unicode-issues).

I pushed an unpolished but for now working [fix](https://github.com/mxmlnkn/ratarmount/commit/98ba888f88453f4991c269761c885cbdccccfe59) to the develop branch. You can try it out with `pip3 install --upgrade git+https://github.com/mxmlnkn/ratarmount.git@develop#egginfo=ratarmount`. The problematic characters will now be escaped but you can try to specify the correct filename encoding with, e.g., `--encoding latin1` to make the filenames look perfect.

---

More in detail deliberations:

It looks like your TAR contains file names in a possibly foreign language, is that possible? The mentioned surrogate character `\udc83` [is basically just Python](https://stackoverflow.com/a/27367173/2191065) trying to bijectively map the unknown input encoding for the character `Ã`(?) to UTF-8. You can arrive at the same exception with this:
```python
b"\x83".decode( 'utf8','surrogateescape' ).encode( 'utf8' )
```
In this case. we could fix this with:
```python
b"\x83".decode( 'iso-8859-1' ).encode( 'utf8' )
```
Or if we already have the decoded name beause that's what tarfile gives me:
```python
tarFileGivenFname.encode( 'utf8', 'surrogateescape' ).decode( 'iso-8859-1' )
```
However, for that you would have to know the correct encoding scheme first. And, I would need to add a new option to specify the tar file name encoding.

I was able to create a tar reproducing the error using the [iconv character conversion module](https://www.systutorials.com/docs/linux/man/8-mount.fuse/) of FUSE itself and the bind mounting ability of ratarmount:
```bash
mkdir mimi momo
touch mimi/Datei-mit-dämlicher-Kodierung.txt
ratarmount -o modules=iconv,to_code=ISO-8859-1 mimi momo
tar -c --owner=0 --group=0 --numeric-owner -f special-char.tar momo/*
fusermount -u momo
ratarmount special-char.tar
```

```python
Traceback (most recent call last):
  File "/home/user/.local/bin/ratarmount", line 8, in <module>
    sys.exit(cli())
  File "/home/user/.local/lib/python3.7/site-packages/ratarmount.py", line 1562, in cli
    mountPoint           = args.mount_point )
  File "/home/user/.local/lib/python3.7/site-packages/ratarmount.py", line 997, in __init__
    for tarFile in pathToMount ]
  File "/home/user/.local/lib/python3.7/site-packages/ratarmount.py", line 997, in <listcomp>
    for tarFile in pathToMount ]
  File "/home/user/.local/lib/python3.7/site-packages/ratarmount.py", line 1043, in _openTar
    gzipSeekPointSpacing = gzipSeekPointSpacing  )
  File "/home/user/.local/lib/python3.7/site-packages/ratarmount.py", line 294, in __init__
    self.createIndex( self.tarFileObject )
  File "/home/user/.local/lib/python3.7/site-packages/ratarmount.py", line 479, in createIndex
    self._setFileInfo( fileInfo )
  File "/home/user/.local/lib/python3.7/site-packages/ratarmount.py", line 658, in _setFileInfo
    ','.join( '?' * len( row ) ) + ');', row )
UnicodeEncodeError: 'utf-8' codec can't encode character '\udce4' in position 11: surrogates not allowed
Exception ignored in: <function TarMount.__del__ at 0x7f210a138dd0>
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.7/site-packages/ratarmount.py", line 1036, in __del__
    os.close( self.mountPointFd )
AttributeError: mountPointFd
```

What I implemented in develop:
 - A command line option for the encoding, which is given to tarfile and used not only for the file names and such but also for the PAX headers, I think. This is independent of FUSE iconv module encoding options.
 - Non-convertible characters are now detected and a warning is printed for each affected filename. The character is then escaped. E.g., `ä` -> `\xe4`.
 - Add minimal problematic TAR and test.

Todo:
 - Check that the encoding is properly propagated to anywhere necessary.
 - Think about how to make it work with union mount. Basically the same issues as in #31. Currently, it's not possible to set the encoding per input, only for all at once.
 - Maybe put the encoding warning behind a `debug >= 1` because the fallback seem to work pretty good.

**Comment by mxmlnkn** [comment]
Fix is in release 0.6.0
